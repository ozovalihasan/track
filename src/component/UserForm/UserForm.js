import PropTypes from 'prop-types';
import styled from 'styled-components';
import Button from '../Button/Button';
import Input from '../Input/Input';
import ProfileContainer from '../Profile/ProfileContainer';

const UserForm = ({
  handleChange,
  handleSubmit,
  username,
  password,
  buttonName,
}) => (
  <Main>
    <div>
      <ProfileContainer />
    </div>
    <FormContainer>
      <Form onSubmit={handleSubmit}>
        <Input
          onChange={e => handleChange(e, 'password')}
          type="text"
          name="username"
          value={username}
          placeholder="Username"
        />
        <Input
          onChange={e => handleChange(e, 'password')}
          type="password"
          name="password"
          value={password}
          placeholder="Password"
        />
        <SubmitButton type="submit">{buttonName}</SubmitButton>
      </Form>
    </FormContainer>
  </Main>
);

UserForm.propTypes = {
  handleChange: PropTypes.func.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  username: PropTypes.string.isRequired,
  password: PropTypes.string.isRequired,
  buttonName: PropTypes.string.isRequired,
};

const Main = styled.div`
  display: block;
`;

const FormContainer = styled.div`
  width: 100%;
  display: flex;
  height: 100%;
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;
const Form = styled.form`
  display: flex;
  height: 100%;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  text-align: center;
  width: 200px;

`;

const SubmitButton = styled(Button)`
  margin: 40px 10px;
  width: 100%;
`;

export default UserForm;
