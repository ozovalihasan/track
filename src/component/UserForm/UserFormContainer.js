import { useState } from 'react';
import { useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import UserForm from './UserForm';

const UserFormContainer = ({ fetchUser, buttonName }) => {
  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');

  const dispatch = useDispatch();
  const handleSubmit = e => {
    e.preventDefault();
    if (username && password) {
      dispatch(fetchUser(username, password));
    }
  };

  const handleChange = e => {
    if (e.target.name === 'username') {
      setUsername(e.target.value);
    } else if (e.target.name === 'password') {
      setPassword(e.target.value);
    }
  };

  return (
    <UserForm
      handleSubmit={handleSubmit}
      handleChange={handleChange}
      username={username}
      password={password}
      buttonName={buttonName}
    />
  );
};

UserFormContainer.propTypes = {
  fetchUser: PropTypes.func.isRequired,
  buttonName: PropTypes.string.isRequired,
};

export default UserFormContainer;
