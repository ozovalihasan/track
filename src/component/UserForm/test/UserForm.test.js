import { render, screen } from '@testing-library/react';
import React from 'react';
import '@testing-library/jest-dom';

import userEvent from '@testing-library/user-event';
import UserForm from '../UserForm';

jest.mock('../../Profile/ProfileContainer', () => {
  const ProfileContainer = () => (<div>Mock ProfileContainer </div>);
  ProfileContainer.displayName = 'ProfileContainer';
  return ProfileContainer;
});

const username = 'Mock Username';
const password = 'Mock Password';
const buttonName = 'Mock Button Name';

const handleChange = jest.fn();
const handleSubmitSpy = jest.fn();
const handleSubmit = e => {
  handleSubmitSpy();
  e.preventDefault();
};
const renderedComponent = () => render(
  <UserForm
    handleChange={handleChange}
    handleSubmit={handleSubmit}
    username={username}
    password={password}
    buttonName={buttonName}
  />,
);

afterEach(() => {
  handleChange.mockClear();
  handleSubmitSpy.mockClear();
});

describe('<UserForm />', () => {
  it('Mock ProfileContainer is present in DOM', () => {
    renderedComponent();

    expect(screen.getByText(/Mock ProfileContainer/i)).toBeInTheDocument();
  });

  it('call handleSubmit when the form is submitted', () => {
    renderedComponent();

    expect(handleSubmitSpy).not.toHaveBeenCalled();

    userEvent.click(screen.getByText(/Mock Button Name/i));

    expect(handleSubmitSpy).toHaveBeenCalled();
  });

  it('call handleChange when input of username is changed ', () => {
    renderedComponent();

    expect(handleChange).not.toHaveBeenCalled();

    userEvent.type(screen.getByDisplayValue(/Mock Username/i), '1');

    expect(handleChange).toHaveBeenCalled();
  });

  it('call handleChange when input of password is changed ', () => {
    renderedComponent();

    expect(handleChange).not.toHaveBeenCalled();

    userEvent.type(screen.getByDisplayValue(/Mock Password/i), '1');

    expect(handleChange).toHaveBeenCalled();
  });

  it('renders correctly', () => {
    expect(renderedComponent()).toMatchSnapshot();
  });
});
