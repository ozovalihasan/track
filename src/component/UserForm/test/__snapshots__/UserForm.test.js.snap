// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserForm /> renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-bdfBwQ iCancN"
      >
        <div>
          <div>
            Mock ProfileContainer 
          </div>
        </div>
        <div
          class="sc-dlfnbm bMORNo"
        >
          <form
            class="sc-hKgILt klxLsr"
          >
            <input
              class="sc-gsTCUz ezFpSP"
              name="username"
              placeholder="Username"
              type="text"
              value="Mock Username"
            />
            <input
              class="sc-gsTCUz ezFpSP"
              name="password"
              placeholder="Password"
              type="password"
              value="Mock Password"
            />
            <button
              class="sc-eCssSg cfYtMe"
              type="submit"
            >
              Mock Button Name
            </button>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-bdfBwQ iCancN"
    >
      <div>
        <div>
          Mock ProfileContainer 
        </div>
      </div>
      <div
        class="sc-dlfnbm bMORNo"
      >
        <form
          class="sc-hKgILt klxLsr"
        >
          <input
            class="sc-gsTCUz ezFpSP"
            name="username"
            placeholder="Username"
            type="text"
            value="Mock Username"
          />
          <input
            class="sc-gsTCUz ezFpSP"
            name="password"
            placeholder="Password"
            type="password"
            value="Mock Password"
          />
          <button
            class="sc-eCssSg cfYtMe"
            type="submit"
          >
            Mock Button Name
          </button>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
